### PR TITLE
[Merged by Bors] - Validate vertex attribute format on insert

### PR DESCRIPTION
**Pull request by IceSentry**
# Objective

- Validate the format of the values with the expected attribute format.
- Currently, if you pass the wrong format, it will crash somewhere unrelated with a very cryptic error message, so it's really hard to debug for beginners.

## Solution

- Compare the format and panic when unexpected format is passed

## Note

- I used a separate `error!()` for a human friendly message because the panic message is very noisy and hard to parse for beginners. I don't mind changing this to only a panic if people prefer that.
- This could potentially be something that runs only in debug mode, but I don't think inserting attributes is done often enough for this to be an issue.

### CODE REVIEW

**Review by afonsolage — COMMENTED** [review]
Related to this discussion: https://github.com/bevyengine/bevy/discussions/5257

**Review by ManevilleF — APPROVED** [review]

**Comment by ManevilleF** [comment]
Could you add a `Panics` section to the function doc strings?

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 1]
```diff
@@ -107,13 +107,19 @@ impl Mesh {
...
-                attribute,
-                values: values.into(),
-            },
-        );
+        let values: VertexAttributeValues = values.into();
+
+        let values_format = VertexFormat::from(&values);
+        if values_format != attribute.format {
+            error!(
+                "Invalid attribute format for {} given format is {:?} but expected {:?}",
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/mesh/mesh/mod.rs:115]:
```suggestion
                "Invalid attribute format for {}. Given format is {:?} but expected {:?}",
```

**Review by alice-i-cecile — APPROVED** [review]

**Review by devil-ira — APPROVED** [review]
I have a wording suggestion, but this looks good to go!

... And github yeeted my suggestion. gimme a sec.

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 2]
```diff
@@ -101,19 +101,28 @@ impl Mesh {
 
     /// Sets the data for a vertex attribute (position, normal etc.). The name will
     /// often be one of the associated constants such as [`Mesh::ATTRIBUTE_POSITION`].
+    ///
+    /// # Panics
+    /// Panics when the format of the values is not the expected attribute format
```
Comment (devil-ira) [inline crates/bevy_render/src/mesh/mesh/mod.rs:106]:
```suggestion
    /// Panics when the format of the values does not match the attribute's format.
```

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 3]
```diff
@@ -101,19 +101,28 @@ impl Mesh {
...
-        self.attributes.insert(
-            attribute.id,
-            MeshAttributeData {
-                attribute,
-                values: values.into(),
-            },
-        );
+        let values: VertexAttributeValues = values.into();
+
+        let values_format = VertexFormat::from(&values);
```
[the remark this answers is not in the document]
Comment (IceSentry) [reply crates/bevy_render/src/mesh/mesh/mod.rs:115]: That's fair, but why not add it to this PR? It would make that code clearer and it seems directly related.

[the remark this answers is not in the document]
Comment (devil-ira) [reply crates/bevy_render/src/mesh/mesh/mod.rs:115]: Ah, I just didn't want to push it onto you, could've worded that differently.
Feel free to :)

**Comment by IceSentry** [comment]
Oops, I deleted the branch by mistake 😅
